blog-posts/Starting-with-typescritp/Starting-with-typescritp.md

```diff
@@ -0,0 +1,201 @@
+
+```ts
+class Dog {
+  name: string;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+
+  eat(): void {
+    console.log('Glump');
+  }
+}
+```
+
+- Interface is like the structure of a class, it do not implement how things work, only define the API that a objects or class are going to have. Here is an example:
+
+```ts
+interface Animal {
+  name: string;
+  eat(): void;
+}
+```
+
+here an example of a class implementing the interface Animal and a function that returns that interface.
+
+```ts
+class Human implements Animal {
+  name: string = 'John';
+
+  eat(): void {
+    console.log('gluh');
+  }
+}
+
+class Bird implements Animal {
+  name: string = 'Parrot';
+
+  eat(): void {
+    console.log('glip');
+  }
+}
+
+function bornAnimal(type: string): Animal {
+  if (type === 'BIRD') {
+    return new Bird();
+  }
+
+  if (type === 'HUMAN') {
+    return new Human();
+  }
+}
+```
+
+So Interfaces let us use objects without knowing exactly how it is implemented or his entire API, only understanding the part that is exposed by the interface.
+
+- Types: It is really similar to an interface, they are used to define the structure of an object but it has some difference, for now lets show how to use it
+
+```ts
+type Animal = {
+  name: string;
+  eat(): void;
+};
+```
+
+if you want to know more about differences about interfaces and types [here is a good post about it](https://medium.com/@martin_hotell/interface-vs-type-alias-in-typescript-2-7-2a8f1777af4c). To keep it simple lets say that its usually used to define literal object and for more compact definitions, for example to define states and props in React with Typescript.
+
+## Modifiers : private, public, protected
+
+The modifiers define how a property of an object can be access. So lets show with examples an short explanations
+
+- public: It can be used by everyone, it is part of the public API of an object.
+- protected: Only classes that extend that object can use this property.
+- private: It is only usable inside the object, not accessible from outside of object not even from subclasses.
+
+```ts
+class Human {
+  public name: string;
+  protected gender: string;
+  private genes: string;
+}
+
+class John extends Human {
+  name = 'John';
+  gender = 'male';
+  genes = 'ABXBSA'; // this is not allowed
+
+  constructor(genes) {
+    this.genes = genes;
+  }
+}
+
+const foo: Human = new John();
+console.log(foo.name);
+console.log(foo.gender); // this is not allowed
+```
+
+## Types composing || and &
+
+# Found a typo?
+
+If you've found a typo, a sentence that could be improved or anything else that should be updated on this blog post, you can access it through a git repository and make a pull request. Instead of posting a comment, please go directly to [my github repository](https://github.com/ramclen/documenting-dev) and open a new pull request with your changes.
```

Comment:
Advantages

- It always highlights errors at compilation time during the time of development.
- 
- Safe automatic refactorings
- Allows great tooling supports with IntelliSense which provides active hints as the code is added.

Disadvantages
- Long effort to incorporate the code.
- TypeScript takes a longer time to compile the code as compared to the JavaScript, because it needs to transform typescript to javascript.
- TypeScript does not support abstract classes.

-------

How do you compile TypeScript files?
The extension for any TypeScript file is .ts.

Reply:
I like those ideas, I will talk about it in next post